### ISSUE TITLE
ped exits immediately with error in Anaconda Python 3.8.*

### ISSUE THREAD

**Issue by rg00dwin** [jpfxgood/ped#5]
Running the following:

`python ped`

The editor exits immediately back to terminal with error:

`byte doesn't fit in chtype`

Fearing that I had inadvertently changed something in the local source, I did the following:
```
Opened up new terminal window
conda activate py38
export SSH_DIALOG_USERNAME=roy;export SSH_DIALOG_PASSWORD=REDACTED;export SSH_DIALOG_BASEPATH=ssh://bigboy/tmp
cd /home/roy/python
mkdir ped_c
cd ped_c
downloaded and unzipped https://github.com/jpfxgood/ped/archive/v1.0.1.tar.gz
export PATH=/home/roy/python/ped_c/ped-1.0.1:$PATH
cd ped-1.0.1
python ped
```

editor exited immediately with 

`byte doesn't fit in chtype`

**Comment by jpfxgood** [comment]
~/ped.log should exist and the traceback should be at the end of that file, I'd be interested to see it
BTW I can't reproduce that locally, but I'm not running 3.8 of python I'm running 3.6.9 so perhaps something became even stricter

**Comment by rg00dwin** [comment]
[ped.log](https://github.com/jpfxgood/ped/files/4873109/ped.log)

I'm running python 3.8 since it was mentioned in the README.md.. 

This might not even be reproducible with the 1.10 release..

**Comment by rg00dwin** [comment]
Reproducible with 1.1.0 
[ped.log](https://github.com/jpfxgood/ped/files/4873133/ped.log)

**Comment by rg00dwin** [comment]
It goes (almost) without saying that I would feel regretful should this issue stem from some configuration problem on my part.. I am going to create a python 3.6.9 conda environment and try it with that..

**Comment by jpfxgood** [comment]
Yeah that is a weird one because curses says that it returns an int not bytes or character for those calls
I'm going to have to set up a 3.8 dev environment and debug that to find out what type is actually coming back from the scr.getch call just before... it is extremely weird and perhaps a bug in the curses module if the output of scr.getch is incompatible with the input curses.ungetch I'll have to go look and see if there are any issues there

**Comment by rg00dwin** [comment]
so - 

if I run ped by itself it exits immediately

But - if I run ped with a filename "ped test.txt" it brings up the editor

**Comment by rg00dwin** [comment]
```
py369) roy@bigboy:~/anaconda3/envs/py369/bin$ python --version
Python 3.6.9 :: Anaconda, Inc.
```

**Comment by jpfxgood** [comment]
what version do you get when you do "python3 --version" ?
ped has a hash bang that uses python3 directly so it'll run on systems where python3 isn't the default

**Comment by rg00dwin** [comment]
```
(py369) roy@bigboy:~/anaconda3/envs/py369/bin$ python3 --version
Python 3.6.9 :: Anaconda, Inc.
(py369) roy@bigboy:~/anaconda3/envs/py369/bin$ 

```

**Comment by rg00dwin** [comment]
```
(py369) roy@bigboy:~/anaconda3/envs/py369/bin$ cd
(py369) roy@bigboy:~$ pwd
/home/roy
(py369) roy@bigboy:~$ python3 --version
Python 3.6.9 :: Anaconda, Inc.
(py369) roy@bigboy:~$ 

```

**Comment by jpfxgood** [comment]
sorry, forgot a part, I need "/usr/bin/env python3 --version"...

**Comment by rg00dwin** [comment]
```
cd /usr/bin
ls -l
<snio>
lrwxrwxrwx 1 root root            9 Apr 16  2018  python -> python2.7
lrwxrwxrwx 1 root root            9 Apr 16  2018  python2 -> python2.7
-rwxr-xr-x 1 root root      3637096 Apr 15 13:20  python2.7
lrwxrwxrwx 1 root root            9 Oct 25  2018  python3 -> python3.6
-rwxr-xr-x 2 root root      4526456 Apr 17 21:56  python3.6
-rwxr-xr-x 2 root root      4526456 Apr 17 21:56  python3.6m
lrwxrwxrwx 1 root root           10 Oct 25  2018  python3m -> python3.6m
</snip>

```

```
py369) roy@bigboy:/usr/bin$ pwd
/usr/bin
(py369) roy@bigboy:/usr/bin$ ./python3 --version
Python 3.6.9
(py369) roy@bigboy:/usr/bin$ ^C
(py369) roy@bigboy:/usr/bin$ 
```

**Comment by rg00dwin** [comment]
```
(py369) roy@bigboy:/usr/bin$ /usr/bin/env python3 --version
Python 3.6.9 :: Anaconda, Inc.
(py369) roy@bigboy:/usr/bin$ 
```

**Comment by rg00dwin** [comment]
gonna try it on my MBP

**Comment by jpfxgood** [comment]
I installed python3.8 on my ubuntu dev machine and made it the current alternative for python3 and I can't reproduce the problem
I did have to force the reinstallation of the dependencies because paramiko barfed but it works fine 
I don't use anaconda, and it looks like they have their own distribution of python, perhaps there's something off about their current curses package?

**Comment by rg00dwin** [comment]
ok cool - I'll get rid of anaconda from bigboy.. and try it again.

**Comment by rg00dwin** [comment]
did you pull from the deadsnakes ppa?

**Comment by jpfxgood** [comment]
nope, ubuntu has a python3.8 package for 18.04 in the repo you can just do apt install python3.8 it won't replace python3.6 so you do: 
sudo apt install python3.8
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.6 10
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 20
sudo update-alternatives --config python3
and on the --config one you choose 3.8 as the alternative to make it current for /usr/bin/python3....

**Comment by jpfxgood** [comment]
I got the tests to pass under python3.8... probably should have done that before I made a new release ;-) I'm a little rusty at this individual contributor gig ;-)

**Comment by jpfxgood** [comment]
my version of vscode does not like python3.8 on ubuntu it was getting an unhandled exception when that was the default python3 so I switched back to 3.6.9... but now I have a way to test using python 3.8 :-)

**Comment by rg00dwin** [comment]
hah - I smoked bigboy...

I uninstalled the 3.6.9 version of python.. 

wow.. you're not supposed to do that.. apparently.. 

https://askubuntu.com/questions/1131731/how-to-remove-python-3-6-from-ubuntu

Anyway - I'm rebuilding him now - laying down 20.04 LTS and the KDE Plasma desktop..  I'll probably be occupied with him for a day or so.. 

bigboy is out in the family room now hardwired to the router.. As you can imagine this is a very upsetting time for him.. all the hubbub out here.. Once I get the GUI running - I'll get his wireless back and then move him back into his snug office..  

The good thing is that  my images are on a different drive from the boot drive (500GB SSD).. 

I might intall windows 10 on him as a dual boot kind of thing so I can run lightroom w/out having to use VBox.. we'll see..

**Comment by jpfxgood** [comment]
Sorry about that 😭 I thought you could install it next to 3.6.9...what distro are you running?

**Comment by rg00dwin** [comment]
I was running bionic beaver (18.04) - am running 20.04 now.. working on getting wireless configured now..

**Comment by jpfxgood** [comment]
So, I installed the individual edition of Anaconda on my laptop Ubuntu 18.04 and created a python 3.8 environment and I couldn't reproduce this with the v1.1.1 release... I even installed it in the conda environment and it worked fine...
I notice that the --version text is different on my install it doesn't say Anaconda,Inc. are you running a paid version?
I'm going to close this since I think it works in all of the open source python 3.8's out there...

**Comment by rg00dwin** [comment]
no - wasn't running a paid version..

**Comment by rg00dwin** [comment]
... of anaconda.

**Comment by rg00dwin** [comment]
An update on this - mostly to convince myself I'm not going insane.

The environment is as close to pristine as one can get - fresh ubuntu 20.04 install on top of a wiped SSD

Installed anaconda 
created a python  3.8.3 environment
pip installed ped-editor

conda version:

```
(py383) roy@bigboy:~/.local/share$ conda --version
conda 4.8.2
(py383) roy@bigboy:~/.local/share$ 
```

contents of the py38 environment:

```
(py383) roy@bigboy:~/python/ped$ conda list
# packages in environment at /home/roy/anaconda3/envs/py383:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
attrs                     19.3.0                     py_0  
bcrypt                    3.1.7            py38h7b6447c_1  
ca-certificates           2020.6.24                     0  
certifi                   2020.6.20                py38_0  
cffi                      1.14.0           py38he30daa8_1  
cryptography              2.9.2            py38h1ba5d50_0  
idna                      2.10                       py_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20191231         h7b6447c_0  
libffi                    3.3                  he6710b0_1  
libgcc-ng                 9.1.0                hdf63c60_0  
libsodium                 1.0.18               h7b6447c_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
more-itertools            8.4.0                      py_0  
ncurses                   6.2                  he6710b0_1  
openssl                   1.1.1g               h7b6447c_0  
packaging                 20.4                       py_0  
paramiko                  2.7.1                      py_0  
ped-editor                1.1.1                    pypi_0    pypi
pip                       20.1.1                   py38_1  
pluggy                    0.13.1                   py38_0  
py                        1.9.0                      py_0  
pycparser                 2.20                       py_0  
pygments                  2.6.1                      py_0  
pynacl                    1.4.0            py38h7b6447c_0  
pyparsing                 2.4.7                      py_0  
pytest                    5.4.3                    py38_0  
python                    3.8.3                hcff3b4d_2  
readline                  8.0                  h7b6447c_0  
setuptools                47.3.1                   py38_0  
six                       1.15.0                     py_0  
sqlite                    3.32.3               h62c20be_0  
tk                        8.6.10               hbc83047_0  
wcwidth                   0.2.5                      py_0  
wheel                     0.34.2                   py38_0  
xz                        5.2.5                h7b6447c_0  
zlib                      1.2.11               h7b6447c_3  

```

went to the directory where pip installed ped to and ran: 

`./ped
`

saw:

```
py383) roy@bigboy:~/anaconda3/envs/py383/bin$ ./ped
byte doesn't fit in chtype
(py383) roy@bigboy:~/anaconda3/envs/py383/bin$ 
```

Contents of ~/.pedbackup/ped.log

```
ERROR:root:byte doesn't fit in chtype
ERROR:root:Traceback (most recent call last):
  File "/home/roy/anaconda3/envs/py383/bin/ped", line 73, in main
    e.openEditor()
  File "/home/roy/anaconda3/envs/py383/lib/python3.8/site-packages/ped_core/editor_manager.py", line 427, in openEditor
    choices = f.main()
  File "/home/roy/anaconda3/envs/py383/lib/python3.8/site-packages/ped_dialog/dialog.py", line 840, in main
    if (not keymap.keypending(self.win)) or force:
  File "/home/roy/anaconda3/envs/py383/lib/python3.8/site-packages/ped_core/keymap.py", line 198, in keypending
    curses.ungetch(ch)
OverflowError: byte doesn't fit in chtype
```

**Comment by jpfxgood** [comment]
Finally reproduced it, and fixed it... the code to make it happen is here:

    import curses

    def main(stdscr):
        curses.raw()
        stdscr.nodelay(1)
        ch = stdscr.getch()
        print("The integer is %d"%ch,file=open("test.out","a"))
        curses.ungetch(ch)

    if __name__ == '__main__':
        curses.wrapper(main)

So the fix will be in the next patch...
I can't reproduce this in any other python3.8 like the direct ubuntu package, it is somehow special for anaconda...

**Comment by jpfxgood** [comment]
From their blog about their release from March 2020:

> As there are still some key OSS projects that aren’t available or stable for Python 3.8, the installers are based around Python 3.7. Notably, at the time of this release, Tensorflow does not yet support Python 3.8, so installers based around Python 3.8 would cause problems for users who want to use Tensorflow. There are Python 3.6 and Python 3.8 metapackages available with this release, so you can work with Anaconda in other versions of Python as you wish.

This would make me think that 3.8 isn't well supported or recommended on their platform and that 3.7 is.

**Comment by rg00dwin** [comment]
Isn't that something? 

I'll make my container with python 3.7

here's ped running in the  python 3.8 container..

![image](https://user-images.githubusercontent.com/18055220/86638687-c3943300-bfa5-11ea-8ab8-99cbbf1b964f.png)

**Comment by rg00dwin** [comment]
when you put up another release - I'll rebuild the container and re-test. 

Also going to make a container that will run the pytest test suite.. 

Fun!

**Comment by jpfxgood** [comment]
I'm thinking of adding a dev_requirements.txt file because now you need: pytest, pdoc, and pexpect ... :-) I guess I should just make a 1.1.2 release...

**Comment by jpfxgood** [comment]
Fixed in 1.1.2
